### PR TITLE
Codec for vtkImageData<->igtl::ImageMessage

### PR DESCRIPTION
**Pull request by christiana**
Split vtk<->igtl conversion from vtkIGTLToMRMLImage into a separate codec class igtlCodecImage.

This is the first step in an attempt to split out all vtk<->igtl code from the OpenIGTLinkIF repo and eventually move it to the OpenIGTLink repo. See http://www.na-mic.org/Wiki/index.php/2016_Winter_Project_Week/Projects/TrackedUltrasoundStandardization for background.

```
Note: this commit compiles on mac but it not yet tested.
```

### CODE REVIEW

**Comment by lassoan** [comment]
Would it make sense to support body compression at OpenIGTLink message level?

Advantage would be that it would be easy to enable compression for other data types. We could add compression type to the extended header.

**Comment by christiana** [comment]
I would be interested in such a compression. A few comments:
- compression speed is important.
- lossy compression (jpeg) is way faster than zlib compression, based on informal tests I did last month. When streaming video (using gstreamer) we ended up using jpeg with a low compression ratio.
- If we are to support jpeg, this will be a special case for the image data only.

A generic lossless algorithm would be great, but we should verify that this is fast enough (given a few use cases) before implementing it.

-Christian

On 10. jan. 2016, at 21.26, Andras Lasso <notifications@github.com<mailto:notifications@github.com>> wrote:

Would it make sense to support body compression at OpenIGTLink message level?

Advantage would be that it would be easy to enable compression for other data types. We could add compression type to the extended header.

**Comment by lassoan** [comment]
I've completely misinterpreted the commit (just by reading the commit message earlier): seeing the "codec" word, I assumed you implemented compression. It would be better to use a more general word, such as converter, adapter, translator (igtlImageTranslator, igtlImageConverter, igtlImageAdapter).

Another naming issue: file name prefix should indicate what kind of class it is. igtl... means that the class is based on igtlObject. As the adapter classes are based on vtkObject, they should be called something like vtkImageTranslator).

Otherwise it looks good, great start, thank you!

File: MRML/vtkIGTLToMRMLImage.cxx [hunk 1]
```diff
@@ -256,303 +219,67 @@ int vtkIGTLToMRMLImage::IGTLToMRML(igtl::MessageBase::Pointer buffer, vtkMRMLNod
     return 0;
     }
 
-  // Create a message buffer to receive image data
-  igtl::ImageMessage::Pointer imgMsg;
-  imgMsg = igtl::ImageMessage::New();
-  imgMsg->Copy(buffer); // !! TODO: copy makes performance issue.
+  if (!Codec.Get())
```
Comment (adamrankin) [inline MRML/vtkIGTLToMRMLImage.cxx]: Minor coding convention comment: typically local variables in functions start with a lower case letter, and member variables start with an uppercase.

File: MRML/vtkIGTLToMRMLImage.cxx [hunk 2]
```diff
@@ -256,303 +219,67 @@ int vtkIGTLToMRMLImage::IGTLToMRML(igtl::MessageBase::Pointer buffer, vtkMRMLNod
...
+    Codec = vtkSmartPointer<igtlCodecImage>::New();
```
Comment (adamrankin) [inline MRML/vtkIGTLToMRMLImage.cxx:223]: Perhaps move this to the constructor and remove from the two converters?

**Comment by adamrankin** [comment]
I agree with Andras regarding the naming. Codec is short for compression/decompression, where as this is strictly a conversion operation.

Great start!

**Comment by jgroehl** [comment]
I also agree on the naming convention as proposed by Andras and Adam. 

Maybe have something like IgtlToVtk_____Converter, where the _____ is in form of Image, PolyData, Transform, TrackingData, etc..?

I would also propose having separate classes for conversion in either direction. This way classes are named exactly the way they behave and the end user can include exactly what he needs. If the same algorithms are needed for both, it may makes sense to extract that in a base class?

Nice work so far :-)

**Comment by christiana** [comment]
Thanks for the replies, all of you. Next time I will try to open an issue so that we can discuss there before the commit. 

According to https://en.wikipedia.org/wiki/Codec, codec=coding/decoding.

Having said that, I agree that Converter is a better name. Far less confusing :-)

Given that the plan is to move the converters to the OpenIGTLink repo, the namespace will be igtl and Class names will be <XXX>Converter, e.g. igtl::ImageConverter, igtl::PolyDataConverter etc. As to @jgroehl´s suggestion of separater to/from-classes, I prefer classes forming a concept and methods executing jobs within that concept, thus keeping them as they are. Please give opinions on this.

@lassoan: The inheritance from vtkObject was simply because the split was performed by copy-paste from the original and then reducing. I suggest we inherit from nothing, and thus we can use naming as suggested above.

@adamrankin: Agreed: move Codec (now Converter) to constructor.

**Comment by lassoan** [comment]
Using a common base class allows a lot of useful features (PrintSelf, copy, more clear and efficient memory management, already existing documentation and conventions, etc), so I would suggest to use an existing base class. If we use igtlLightObject then you the namespace and the base class would be the same, so the file name prefix would be trivial choice (igtl).

Separating two conversion directions to different classes: These converters are already very small (most of the time), there are many of them, they must be registered in pair (it may be more complex and could lead to inconsistencies if converters into different directions could be registered separately), must be fully consistent (if one is changed then most likely the other has to be changed, too), and they often share structures (for example, header structs), utility functions, and description (name of the OpenIGTLink message type that it converts to/from, version of the converter plugin, etc). So, I would suggest to keep them as they are - one converter into both directions as one class.

Of course, all these things are suggestions only - ultimately the person who spend time with developments will make the decision.

**Comment by adamrankin** [comment]
According to https://en.wikipedia.org/wiki/Codec, codec=coding/decoding.

Well I'll be a horse uncle... My apologies.

Adam

**Pull request merged**